ops/relation.py

```diff
@@ -0,0 +1,548 @@
+    charms, with charms that consume those resources. When a consumer
+    charm joins the relation a provider object informs it of the type
+    and version of the resource provided. Providers may also be
+    instrumented to inform consumer charms of any relavent
+    information, for example configuration settings. This is done
+    using the `provides` argument of the :class:`Provider` constructor.
+
+    Any charm that provides a resource may choose to do so through the
+    :class:`Provider` object simpliy by instantiating it in the charm's
+    `__init__` method, as follows
+
+    Example::
+
+        self.provider = Provider(self, relation_name, provides)
+        ...
+        self.provider.ready()
+
+    It is important to invoke `ready()` on the :class:`Provider`
+    object, in order to let the consumer charm know that the provider
+    is serving requests. This is done by setting a boolean flag
+    `ready` in the data forwarded to the consumer charm. A provider
+    charm may toggle this flag by invoking `unready()` when it
+    is unable to service any requests for example during an
+    upgrade. After an upgrade the :class:`Provider` object notifies
+    consumer charms by resending type, version and any other data to
+    the consumer. Even though :class:`Provider` objects only handle
+    relation joined and provider upgrade events, they may be
+    subclassed to extend their functionality in any way desired.
+
+    Args:
+        charm: :class:`ops.charm.CharmBase` derived object that is
+            instantiating :class:`Provider`. This is almost always
+            `self`.
+        name: string name of relation (as defined in `metadata.yaml`) that
+            consumer charms will use to interact with provider charms.
+        provides: A dictionary containing the information that
+            :class:`Provider` will forward to any consumer charm when it forms
+            a relation with the provider charm. This dictionary must have at
+            least one key `provides`. The value of this key itself must be a
+            dictionary, whose key as a resource name and value is a version
+            string. For example here is a suitable provides argument
+            Example::
+
+                {'provides': {'mongodb': '2.4'}}
+
+    The `provides` argument dictionary may contain any other key value
+    pairs. If so this information is forwarded to the consumer charm
+    as application data.
+
+    """
+    stored = StoredState()
+
+    def __init__(self, charm, name, provides):
+        super().__init__(charm, name)
+
+        self.stored.set_default(ready=False)
+        self.name = name
+        self.provides = provides
+
+        events = charm.on[name]
+        self.framework.observe(events.relation_joined, self.on_consumer_joined)
+        self.framework.observe(charm.on.upgrade_charm, self.on_upgrade)
+
+    def on_consumer_joined(self, event):
```

Comment:
These methods also traditionally are class private, since they are responding to events, but shouldn't be called by user code.

Reply:
Ok. Will add underscore prefix.

Reply:
Done.